### ISSUE TITLE
After upgrade from Ubuntu 14 LTS to 16 LTS "Error Connecting" for every game-server

### ISSUE THREAD

**Issue by Mecha-Weasel** [sbpp/sourcebans-pp#228]
I just upgraded my host from Ubuntu 14 LTS to Ubuntu 16 LTS, and am consequently troubleshooting some things.  I ran into and fixed the issue with the upgrade killing my MySQL and PHP installations. Fixed all that, and got Apache to work with PHP (7.0), MySQL server/client stuff upgraded, etc.

## Expected Behavior
Server status updates.

## Current Behavior
So, at this point, the SourceBans++ site now comes-up and does not show any errors accessing the database, etc. But, what does not seem to work any more is having it actually query the game-server instances.

They are running on the same VPS host. But each of them now indicates:

"Error connecting (serverip:serverport)"

PS: Also already tried disabling local firewall on the server, variations on server connection (FQDN, direct public IP, localhost, loop-back IP, etc.). No differences.

## Possible Solution
IDK

## Steps to Reproduce (for bugs)
Live site is:  http://sourcebanspp.weaselslair.com

## Context
Dashboard server status does not upgrade.
Unsure of any other impacts (i.e. are bans being enforced?).

## Your Environment
* Apache: 2.4.18
* PHP: 7.0
* MySQL: 5.7.17 
* Operating System: Ubuntu LTS 16.04.02 (Kernel Linux 4.4.0-64-generic on x86_64)
* Link to your project:  http://sourcebanspp.weaselslair.com

**Comment by Technoblazed** [comment]
Are the servers actually turned on? None of them ping.

**Comment by Mecha-Weasel** [comment]
Yes, 1 is currently on - games.weaselslair.com:6301 = TF2.

Haven't bothered to turn the others on until done troubleshooting.

**Comment by Mecha-Weasel** [comment]
PS: Also upgraded to "dev" version.  But, same issue continues.

**Comment by Mecha-Weasel** [comment]
Did some additional troubleshooting.  Following the information from the FAQ (https://sbpp.github.io/faq/), I edited [b]sb_debug_connection.php[/b] and tested one of the servers (my main TF2 server in this case).

Below are the results.  Keep in mind that I have everything outbound open from the server and even turned-off my firewall entirely anyway for this testing - AND this is all on the same server.

```
[+] SourceBans "Error Connecting()" Debug starting for server 192.223.29.234:6301

[+] Trying to establish UDP connection
[+] UDP connection successfull!
[+] Trying to write to the socket
[+] Successfully requested server info. (That doesn't mean anything on an UDP stream.) Reading...
[-] Error getting server info. Can't read from UDP stream. Port is possibly blocked.

[+] Trying to establish TCP connection
[+] TCP connection successfull!
[+] Trying to write to TCP socket and authenticate via rcon
[+] Successfully sent authentication request. Reading...
[+] Got an response! 
[+] Password correct!
```

**Comment by Groruk** [comment]
Your problem has to be in your webserver configuration since I can get an udp response from your server.

```
[+] SourceBans "Error Connecting()" Debug starting for server 192.223.29.234:6301

[+] Trying to establish UDP connection
[+] UDP connection successfull!
[+] Trying to write to the socket
[+] Successfully requested server info. (That doesn't mean anything on an UDP stream.) Reading...
[+] Got an response! Server: Weasels TF2#1 (Orange): Skins,Critz,RTD,Instaspawn,Music 

[+] Trying to establish TCP connection
[+] TCP connection successfull!
[o] Stopping here since no rcon password specified.
```

Maybe @galexrt can help you with this.

**Comment by galexrt** [comment]
@Mecha-Weasel Can you post the output of the following command `iptables-save`.

Edit: Just saw you already wrote that you are hosting on the same VPS.

**Comment by Mecha-Weasel** [comment]
Yes. Apache, MySQL and all the SRCDS instances are running on the same VPS.

On Feb 25, 2017 2:20 AM, "Alexander Trost" <notifications@github.com> wrote:

@Mecha-Weasel <https://github.com/Mecha-Weasel> do you host SourceBans?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/sbpp/sourcebans-pp/issues/228#issuecomment-282474872>,
or mute the thread

**Comment by galexrt** [comment]
As already written in the comment:
> Can you post the output of the following command iptables-save.

**Comment by Mecha-Weasel** [comment]
NOTE: GitHub seems to be reformatting some of this, but here's the output .. also default action for output is obviously to "accept".
```
# Generated by iptables-save v1.6.0 on Tue Feb 28 20:57:04 2017
*nat
:PREROUTING ACCEPT [1616280:85174655]
:INPUT ACCEPT [1601546:84243709]
:OUTPUT ACCEPT [449571:31399321]
:POSTROUTING ACCEPT [449571:31399321]
COMMIT
# Completed on Tue Feb 28 20:57:04 2017
# Generated by iptables-save v1.6.0 on Tue Feb 28 20:57:04 2017
*mangle
:PREROUTING ACCEPT [194099181:18150673610]
:INPUT ACCEPT [194099181:18150673610]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [196916225:147273598017]
:POSTROUTING ACCEPT [196917232:147273688131]
COMMIT
# Completed on Tue Feb 28 20:57:04 2017
# Generated by iptables-save v1.6.0 on Tue Feb 28 20:57:04 2017
*filter
:INPUT DROP [373514:25956427]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [352458:23130531]
:fail2ban-default - [0:0]
-A INPUT -j fail2ban-default
-A INPUT -s 8.8.8.8/32 -j ACCEPT
-A INPUT -s 8.8.4.4/32 -j ACCEPT
-A INPUT -s 127.0.0.1/32 -i lo -j ACCEPT
-A INPUT -s 192.223.29.234/32 -d 192.223.29.234/32 -i eth0 -j ACCEPT
-A INPUT -p tcp -m tcp -m state --state ESTABLISHED -j ACCEPT
-A INPUT -p icmp -m icmp --icmp-type 8 -j ACCEPT
-A INPUT -p icmp -m icmp --icmp-type 0 -j ACCEPT
-A INPUT -p udp -m udp --sport 53 --dport 53 -j ACCEPT
-A INPUT -p udp -m udp --sport 123 --dport 123 -j ACCEPT
-A INPUT -p tcp -m tcp --dport 22 -m state --state NEW -j ACCEPT
-A INPUT -p tcp -m tcp --dport 80 -m state --state NEW -j ACCEPT
-A INPUT -p tcp -m tcp --dport 443 -m state --state NEW -j ACCEPT
-A INPUT -s 192.223.29.201/32 -j ACCEPT
-A INPUT -s 103.192.177.108/32 -j ACCEPT
-A INPUT -p udp -m udp -m multiport --dports 6100:6499,7100:7499,8100:8499,9100:9499 -j ACCEPT
-A INPUT -p tcp -m tcp -m multiport --dports 6100:6499,7100:7499,8100:8499,9100:9499 -m state --state NEW -j ACCEPT
-A OUTPUT -d 8.8.8.8/32 -j ACCEPT
-A OUTPUT -d 8.8.4.4/32 -j ACCEPT
-A OUTPUT -m state --state ESTABLISHED -j ACCEPT
-A OUTPUT -p icmp -m icmp --icmp-type 8 -j ACCEPT
-A OUTPUT -p icmp -m icmp --icmp-type 0 -j ACCEPT
-A OUTPUT -d 192.223.29.201/32 -j ACCEPT
-A OUTPUT -d 103.192.177.108/32 -j ACCEPT
-A OUTPUT -p udp -m udp -m multiport --sports 6100:6499,7100:7499,8100:8499,9100:9499 -j ACCEPT
-A OUTPUT -p tcp -m tcp -m multiport --sports 6100:6499,7100:7499,8100:8499,9100:9499 -m state --state NEW -j ACCEPT
-A fail2ban-default -j RETURN
COMMIT
# Completed on Tue Feb 28 20:57:04 2017
```

**Comment by galexrt** [comment]
BTW @Groruk Same debug conn result for me.
```
$ php sb_test.php
[+] SourceBans 'Error Connecting()' Debug starting for server 192.223.29.234:6301

[+] Trying to establish UDP connection
[+] UDP connection successfull!
[+] Trying to write to the socket
[+] Successfully requested server info. (That doesn"t mean anything on an UDP stream.) Reading...
[+] Got an response! Server: Weasels TF2#1 (Orange): Skins,Critz,RTD,Instaspawn,Music
[...]
```

**Comment by galexrt** [comment]
**This is not a problem with SourceBans Web or anyother SourceBans component.**
This is fully a problem related to your setup!

***

I'll be nice with you and tell you what the problem is (aka free "IT support")..
Your problem is your badly made firewall (I guess it is copy and pasted from somewhere..).

What protocol does a UDP socket "respond" in? _In UDP._
What protocol does not have any state to it? _UDP._
What awesome *nix Kernel module allows for conntracking UDP "connections"? _conntrack module._

**And now to your firewall**:
Remove the tcp stuff from the one `INPUT` rule, below is the line and it will begin to work:
```
-A INPUT -p tcp -m tcp -m state --state ESTABLISHED -j ACCEPT
```
Change it to:
```
-A INPUT -m state --state ESTABLISHED -j ACCEPT
```

Change the OUTPUT rule from only allowing `ESTABLISHED` to `NEW,RELATED,ESTABLISHED` or else you'll have other network trouble in the future.
It should be:
```
-A OUTPUT -m state --state NEW,RELATED,ESTABLISHED -j ACCEPT
```

**Comment by Mecha-Weasel** [comment]
What did clear things up for me, was @galexrt's assertion that the rule about allowing "established" connections was in some way too specific (i.e TCP protocol + TCP "established" state).

I applied the change to match what @galexrt suggested, and everything started working as expected.

That rule has worked fine for a long time on this server prior to the in-place-upgrade from Ubuntu 14.04 LTS to Ubuntu 16.04 LTS.  So, I can only conclude there is some change in the underlying behavior of IPTables between the two versions.

What's weird, is I've been doing rules like that on IPTables for like, a decade+- and never had this symptom. LOL.  But thanks to @galexrt for the extra set of eyes and fresh perspective!

Although not a total noob regarding "IT" (only been doing "IT for 30 years, and working on firewalls in various forms since in the mid-1990's, and dedicated game-servers since pre-Y2K).  But, just goes to show a fresh set of eyes on a problem can make all the difference!  I definitely would NOT have thought to do that - given it had been working prior to the upgrade.

For clarity: I did eventually remove all the outbound rules, as the default outbound action was set to "allow" anyway.  They were left-overs from trying (a long time ago) to restrict things - to just ports known to be used by expected processes.  But, unfortunately, there's lots of stuff on the server that only uses dynamically-assigned high-number ports for out-bound and can not be easily be restricted to specific ports. So, none of that ever really worked and I had to set outbound to "allow" by default. anyway So, all those out-bound rules were "moot".
